src/main/scala/is/hail/expr/Relational.scala

```diff
@@ -0,0 +1,340 @@
+package is.hail.expr
+
+import java.io.FileNotFoundException
+
+import is.hail.HailContext
+import is.hail.annotations.Annotation
+import is.hail.methods.{Aggregators, Filter}
+import is.hail.sparkextras.{OrderedPartitioner, OrderedRDD}
+import is.hail.variant.{Genotype, Locus, VSMLocalValue, VSMMetadata, Variant, VariantDataset}
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.Row
+import is.hail.utils._
+import org.apache.spark.SparkContext
+import org.apache.spark.broadcast.Broadcast
+import org.json4s.jackson.JsonMethods
+
+import scala.reflect.ClassTag
+
+case class MatrixType(
+  metadata: VSMMetadata) {
+  def globalType: Type = metadata.globalSignature
+
+  def saType: Type = metadata.saSignature
+
+  def vaType: Type = metadata.vaSignature
+
+  def genotypeType: Type = metadata.genotypeSignature
+
+  def typ = TStruct(
+    "v" -> TVariant,
+    "va" -> vaType,
+    "s" -> TString,
+    "sa" -> saType,
+    "g" -> genotypeType)
+
+  def sampleEC: EvalContext = {
+    val aggregationST = Map(
+      "global" -> (0, globalType),
+      "s" -> (1, TString),
+      "sa" -> (2, saType),
+      "g" -> (3, genotypeType),
+      "v" -> (4, TVariant),
+      "va" -> (5, vaType))
+    EvalContext(Map(
+      "global" -> (0, globalType),
+      "s" -> (1, TString),
+      "sa" -> (2, saType),
+      "gs" -> (3, TAggregable(genotypeType, aggregationST))))
+  }
+
+  def variantEC: EvalContext = {
+    val aggregationST = Map(
+      "global" -> (0, globalType),
+      "v" -> (1, TVariant),
+      "va" -> (2, vaType),
+      "g" -> (3, genotypeType),
+      "s" -> (4, TString),
+      "sa" -> (5, saType))
+    EvalContext(Map(
+      "global" -> (0, globalType),
+      "v" -> (1, TVariant),
+      "va" -> (2, vaType),
+      "gs" -> (3, TAggregable(genotypeType, aggregationST))))
+  }
+}
+
+object NewAST {
+  def genericRewriteTopDown(ast: NewAST, rule: PartialFunction[NewAST, NewAST]): NewAST = {
+    def rewrite(ast: NewAST): NewAST = {
+      rule.lift(ast) match {
+        // FIXME structural compare?
+        case Some(newAST) if newAST ne ast =>
```

Comment:
This is required (err, this, or that rule^(x) = x), otherwise it will go into an infinite loop.  Do you want this realized in the code, e.g., I pass in the n and throw an exception if we iterate too many times?  That works here, but I don't think it works in general because the recursion might be inductive on the input and not the rules, e.g., x + 0 => x requires unbounded iterations depending on the shape of the input.